src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,450 @@
+    if not rules:
+        msg = f"inactive task {uuid}, container: {pname} for Rucio account {rucioAccount} does not have any existing rules, proceed without update"
+        logger.info(msg)
+
+    # persist an up-to-date version of the pileup data structure in MongoDB
+    if modify:
+        logger.info(f"inactive task {uuid}, update {pname}")
+        mgr.updatePileup(doc)
+        msg = f"update pileup {pname}"
+        report.addEntry('inactive', uuid, msg)
+
+
+def activeTask(doc, spec):
+    """
+    Perform single active task over provided MSPileup document
+
+    :param mgr: MSPileupData manager
+    :param doc: MSPileup document
+    :param spec: task spec dict
+    """
+    mgr = spec['manager']
+    uuid = spec['uuid']
+    logger = spec.get('logger', getMSLogger(False))
+    rucioClient = spec['rucioClient']
+    rucioAccount = spec['rucioAccount']
+    report = spec['report']
+    marginSpace = spec.get('marginSpace', 0)
+
+    # extract relevant part of our pileup document we'll use in our logic
+    expectedRSEs = set(doc['expectedRSEs'])
+    currentRSEs = set(doc['currentRSEs'])
+    pileupSize = doc['pileupSize']
+    pname = doc['pileupName']
+    inputArgs = {'scope': 'cms', 'name': pname, 'account': rucioAccount}
+    modify = False
+    localCopyOfCurrentRSEs = list(currentRSEs)
+
+    # if expectedRSEs is different than currentRSEs, then further data placement is required
+    # it's possible that data removal is required!
+    if expectedRSEs != currentRSEs:
+        msg = f"active task {uuid}, further data placement required for pileup name: {pname}, with expectedRSEs: {expectedRSEs} and data currently available at: {currentRSEs}"
+        logger.warning(msg)
+
+        # get list of replication rules for our scope, pileup name and account
+        if isinstance(rucioClient, Client):
+            rules = [r for r in rucioClient.list_replication_rules(inputArgs)]
+        else:
+            rules = [r for r in rucioClient.listDataRules(pname)]
+        # for each rse_expression in rules get list of rses
+        for rdoc in rules:
+            if isinstance(rucioClient, Client):
+                rses = [r['rse'] for r in rucioClient.list_rses(rdoc['rse_expression'])]
+            else:
+                rses = [rse for rse in rucioClient.evaluateRSEExpression(rdoc['rse_expression'])]
+            rdoc['rses'] = rses
+
+        # for each rule matching the DID + Rucio account
+        for rdoc in rules:
+            rid = rdoc['id']
+            rses = rdoc['rses']
+            for rse in rses:
+                # if rule RSE is not in expectedRSEs, then this rule needs to be deleted
+                if rse not in expectedRSEs:
+                    # upon successful rule deletion, also remove the RSE name from currentRSEs
+                    if isinstance(rucioClient, Client):
+                        rucioClient.delete_replication_rule(rid)
+                    else:
+                        rucioClient.deleteRule(rid)
```

Comment:
I am not sure how relevant it is, but should we check the response? Same comment applies to the client API call above.

Reply:
if we should check the response then you need to provide instructions in requirements what to do with this response. For instance, if is clear that if response is ok, but it is unclear if response is not ok. So, if you want me to check it please provide set of actions how to deal with it.

Reply:
we agree to keep it as it since if it fails it is in a log